### ISSUE TITLE
Catboost produces different AUCs on a training set during training and inference

### ISSUE THREAD

**Issue by dlihhats** [catboost/catboost#296]
Problem: Catboost produces different AUCs on a training set during training and inference
catboost version: Python package for catboost version 0.7
Operating System: Ubuntu 16.04.3 LTS
CPU: CPU

I've been using CatboostClassifier with the `eval_set` option to prevent overfitting, and it seemed to be working right according to logs that are output with the `logging_level='Verbose'`.

However, whenever I built the precision-recall curve for the training set, it looked too much different to what I'd expect for the given training AUC in the logs.

So I decided to see what happens if I put the training data into `eval_set` to see the logs. Surprisingly, `learn` AUC was different to `test` AUC.

It's easy to reproduce this with the following code. Note that I'm using the same data for training and for `eval_set`.

   ```
import numpy as np
from catboost import CatBoostClassifier
# initialize data
np.random.seed(42)
train_data = np.random.randint(0, 100, size=(100, 10))
train_label = np.random.randint(0, 2, size=(100))
test_data = np.random.randint(0, 100, size=(50, 10))
test_label = np.random.randint(0, 2, size=(50))
# specify the training parameters 
model = CatBoostClassifier(iterations=20, depth=2, 
                                             learning_rate=1, loss_function='Logloss', 
                                             logging_level='Verbose', eval_metric='AUC', 
                                             random_seed=42)
#train the model
model.fit(train_data, train_label, cat_features=[0,2,5], 
              eval_set=(train_data, train_label))

0:	learn: 0.6557971	test: 0.6557971	best: 0.6557971 (0)	total: 13.5ms	remaining: 256ms
1:	learn: 0.7648953	test: 0.8590982	best: 0.8590982 (1)	total: 24.5ms	remaining: 220ms
2:	learn: 0.8001208	test: 0.8393720	best: 0.8590982 (1)	total: 34ms	remaining: 193ms
3:	learn: 0.8230676	test: 0.8572866	best: 0.8590982 (1)	total: 44ms	remaining: 176ms
4:	learn: 0.8810386	test: 0.8780193	best: 0.8780193 (4)	total: 54.1ms	remaining: 162ms
5:	learn: 0.8884863	test: 0.8824477	best: 0.8824477 (5)	total: 65.6ms	remaining: 153ms
6:	learn: 0.9015700	test: 0.8947262	best: 0.8947262 (6)	total: 75.9ms	remaining: 141ms
7:	learn: 0.9003623	test: 0.8919082	best: 0.8947262 (6)	total: 87.1ms	remaining: 131ms
8:	learn: 0.9297504	test: 0.9035829	best: 0.9035829 (8)	total: 98.1ms	remaining: 120ms
9:	learn: 0.9333736	test: 0.9064010	best: 0.9064010 (9)	total: 107ms	remaining: 107ms
10:	learn: 0.9500805	test: 0.9313607	best: 0.9313607 (10)	total: 117ms	remaining: 95.9ms
11:	learn: 0.9516908	test: 0.9337762	best: 0.9337762 (11)	total: 129ms	remaining: 85.7ms
12:	learn: 0.9611514	test: 0.9371981	best: 0.9371981 (12)	total: 139ms	remaining: 74.6ms
13:	learn: 0.9631643	test: 0.9428341	best: 0.9428341 (13)	total: 150ms	remaining: 64.1ms
14:	learn: 0.9663849	test: 0.9456522	best: 0.9456522 (14)	total: 161ms	remaining: 53.6ms
15:	learn: 0.9688003	test: 0.9468599	best: 0.9468599 (15)	total: 170ms	remaining: 42.5ms
16:	learn: 0.9675926	test: 0.9492754	best: 0.9492754 (16)	total: 181ms	remaining: 31.9ms
17:	learn: 0.9716184	test: 0.9537037	best: 0.9537037 (17)	total: 193ms	remaining: 21.4ms
18:	learn: 0.9736312	test: 0.9541063	best: 0.9541063 (18)	total: 222ms	remaining: 11.7ms
19:	learn: 0.9752415	test: 0.9561192	best: 0.9561192 (19)	total: 233ms	remaining: 0us

bestTest = 0.9561191626
bestIteration = 19

Shrink model to first 20 iterations.
```

**Comment by Noxoomo** [comment]
Hi!

It's OK for learning with categorical features.
CatBoost use special technique to reduce overfitting while dealing with categorical features (see docs for example https://tech.yandex.com/catboost/doc/dg/concepts/algorithm-main-stages_cat-to-numberic-docpage/#algorithm-main-stages_cat-to-numberic). CatBoost replace cat features with estimators of success rate, this estimators in final model are computed from full dataset. But during learning we use trick inspired by online learning and our estimators for objects with index i doesn't use labels from objects i+1… n. 
When we apply the same model to learn set again, we don't known, that this set was used as a train, so we obtained more biased result (because of overfitting)

**Comment by dlihhats** [comment]
@Noxoomo Interesting, thank you very much for the explanation.

However, what if I want to stack together catboost, xgboost, and maybe some other algorithms using their probability predictions in a linear model? I need to make predictions on the training set at first, right? So if during inference xgboost doesn't overfit on the training set but catboost's predictions on the **training** set are just completely overfit because of what you explained above, stacking just doesn't work.

Is there no way to obtain "unbiased" predictions on the training set? I'll understand if you say no :)

**Comment by dlihhats** [comment]
Thanks, @Noxoomo, I'll try that then. This should have been a StackOverflow question instead, I guess. My apologies. Let's close this issue. Many thanks!
